File: src/main/java/engineer/carrot/warren/warren/ssl/WrappedSSLSocketFactory.java
package engineer.carrot.warren.warren.ssl;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.annotation.Nullable;
import javax.net.ssl.*;
import java.io.IOException;
import java.net.InetAddress;
import java.net.Socket;
import java.net.UnknownHostException;
import java.security.KeyManagementException;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

public class WrappedSSLSocketFactory extends SSLSocketFactory {
    final Logger LOGGER = LoggerFactory.getLogger(WrappedSSLSocketFactory.class);

    private SSLSocketFactory factory;

    public WrappedSSLSocketFactory() {
        this.factory = (SSLSocketFactory) SSLSocketFactory.getDefault();
    }

    public WrappedSSLSocketFactory forciblyAcceptCertificatesWithSHA1Fingerprints(Set<String> forciblyAcceptedFingerprintSHA1s) {
        // TODO: check well-formedness of SHA1 strings
        //  Should be: upper case, no punctuation

        SSLSocketFactory socketFactory = this.createFingerprintsAcceptingSocketFactory(forciblyAcceptedFingerprintSHA1s);
        if (socketFactory != null) {
            this.factory = socketFactory;
        }

        return this;
    }

    @Nullable
    private SSLSocketFactory createFingerprintsAcceptingSocketFactory(Set<String> forciblyAcceptedFingerprintSHA1s) {
        TrustManager[] tm = new TrustManager[]{new FingerprintX509TrustManager(forciblyAcceptedFingerprintSHA1s)};

        SSLContext context;
        try {
            context = SSLContext.getInstance("TLS");
        } catch (NoSuchAlgorithmException e) {
            LOGGER.error("Failed to create custom SSL Context for CustomSSLSocketFactory: {}", e);

            return null;
        }

        try {
            context.init(new KeyManager[0], tm, new SecureRandom());
        } catch (KeyManagementException e) {
            LOGGER.error("Failed to initialise custom SSL Context for CustomSSLSocketFactory: {}", e);

            return null;
        }

        return context.getSocketFactory();
    }

    public SSLSocket disableDHEKeyExchange(Socket socket) {
        SSLSocket sslSocket = (SSLSocket) socket;

        List<String> cipherSubset = new ArrayList<>();
        for (String cipherSuites : sslSocket.getEnabledCipherSuites()) {
            if (!cipherSuites.contains("_DHE_")) {
                cipherSubset.add(cipherSuites);
            }
        }

        sslSocket.setEnabledCipherSuites(cipherSubset.toArray(new String[cipherSubset.size()]));

        return sslSocket;
    }

    // SSLSocketFactory

    @Override
    public String[] getDefaultCipherSuites() {
        return this.factory.getDefaultCipherSuites();
    }

    @Override
    public String[] getSupportedCipherSuites() {
        return this.factory.getSupportedCipherSuites();
    }

    @Override
    public Socket createSocket(Socket socket, String host, int port, boolean autoClose) throws IOException {
        return this.factory.createSocket(socket, host, port, autoClose);
    }

    @Override
    public Socket createSocket(String host, int port) throws IOException, UnknownHostException {
        return this.factory.createSocket(host, port);
    }

    @Override
    public Socket createSocket(String host, int port, InetAddress localAddress, int localPort) throws IOException, UnknownHostException {
        return this.factory.createSocket(host, port, localAddress, localPort);
    }

    @Override
    public Socket createSocket(InetAddress localAddress, int localPort) throws IOException {
        return this.factory.createSocket(localAddress, localPort);
    }

    @Override
    public Socket createSocket(InetAddress address, int port, InetAddress localAddress, int localPort) throws IOException {
        return this.factory.createSocket(address, port, localAddress, localPort);
    }
}

File: src/main/kotlin/engineer/carrot/warren/warren/ssl/WrappedSSLSocketFactory.kt
package engineer.carrot.warren.warren.ssl

import engineer.carrot.warren.warren.loggerFor
import javax.net.ssl.*
import java.security.KeyManagementException
import java.security.KeyStore
import java.security.NoSuchAlgorithmException
import java.security.SecureRandom

class WrappedSSLSocketFactory(private val fingerprints: Set<String>?) {
    val LOGGER = loggerFor<WrappedSSLSocketFactory>()

    private fun createDangerZoneSocketFactory(): SSLSocketFactory {
        val tm = arrayOf<TrustManager>(DangerZoneTrustAllX509TrustManager())

        val context: SSLContext
        try {
            context = SSLContext.getInstance("TLS")
        } catch (e: NoSuchAlgorithmException) {
            throw RuntimeException("failed to create danger zone SSL Context for WrappedSSLSocketFactory: {}", e)
        }

        try {
            context.init(arrayOfNulls<KeyManager>(0), tm, SecureRandom())
        } catch (e: KeyManagementException) {
            throw RuntimeException("failed to initialise danger zone custom SSL Context for WrappedSSLSocketFactory: {}", e)
        }

        return context.socketFactory
    }

    private fun createFingerprintsSocketFactory(fingerprints: Set<String>): SSLSocketFactory {
        val tm = arrayOf<TrustManager>(SHA256SignaturesX509TrustManager(fingerprints))

        val context: SSLContext
        try {
            context = SSLContext.getInstance("TLS")
        } catch (e: NoSuchAlgorithmException) {
            throw RuntimeException("failed to create fingerprints SSL Context for WrappedSSLSocketFactory: {}", e)
        }

        try {
            context.init(arrayOfNulls<KeyManager>(0), tm, SecureRandom())
        } catch (e: KeyManagementException) {
            throw RuntimeException("failed to initialise fingerprints custom SSL Context for WrappedSSLSocketFactory: {}", e)
        }

        return context.socketFactory
    }

    fun createDefaultFactory(sslParameters: SSLParameters, context: SSLContext): SSLSocketFactory {
        val trustManagerFactory = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm())
        val keyStore: KeyStore? = null
        trustManagerFactory.init(keyStore)

        context.init(arrayOfNulls<KeyManager>(0), trustManagerFactory.trustManagers, SecureRandom())

        context.supportedSSLParameters.endpointIdentificationAlgorithm = sslParameters.endpointIdentificationAlgorithm
        context.supportedSSLParameters.cipherSuites = sslParameters.cipherSuites
        context.defaultSSLParameters.endpointIdentificationAlgorithm = sslParameters.endpointIdentificationAlgorithm
        context.defaultSSLParameters.cipherSuites = sslParameters.cipherSuites

        return context.socketFactory
    }

    fun createSocket(host: String, port: Int): SSLSocket {
        val context = SSLContext.getInstance("TLS")
        val sslParameters = SSLParameters()
        sslParameters.cipherSuites = SSLContext.getDefault().defaultSSLParameters.cipherSuites.filterNot { it.contains("_DHE_") }.toTypedArray()

        val factory: SSLSocketFactory = if (fingerprints == null) {
            createDefaultFactory(sslParameters, context)
        } else if (fingerprints.isEmpty()) {
            createDangerZoneSocketFactory()
        } else {
            createFingerprintsSocketFactory(fingerprints)
        }

        val socket = factory.createSocket(host, port) as SSLSocket
        socket.sslParameters = sslParameters

        return socket
    }

}
